executor/pipeline/put.rs

```diff
@@ -0,0 +1,160 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::{collections::HashMap, sync::Arc};
+
+use answer::variable_value::VariableValue;
+use compiler::{
+    executable::{function::ExecutableFunctionRegistry, insert::VariableSource, put::PutExecutable},
+    VariablePosition,
+};
+use lending_iterator::LendingIterator;
+use storage::snapshot::WritableSnapshot;
+
+use crate::{
+    batch::Batch,
+    match_executor::MatchExecutor,
+    pipeline::{
+        stage::{ExecutionContext, StageAPI},
+        PipelineExecutionError, WrittenRowsIterator,
+    },
+    row::MaybeOwnedRow,
+    ExecutionInterrupt,
+};
+
+pub struct PutStageExecutor<PreviousStage> {
+    executable: Arc<PutExecutable>,
+    previous: PreviousStage,
+    function_registry: Arc<ExecutableFunctionRegistry>,
+}
+
+impl<PreviousStage> PutStageExecutor<PreviousStage> {
+    pub fn new(
+        executable: Arc<PutExecutable>,
+        previous: PreviousStage,
+        function_registry: Arc<ExecutableFunctionRegistry>,
+    ) -> Self {
+        Self { executable, previous, function_registry }
+    }
+
+    pub(crate) fn output_width(&self) -> usize {
+        self.executable.output_width()
+    }
+}
+
+impl<Snapshot, PreviousStage> StageAPI<Snapshot> for PutStageExecutor<PreviousStage>
+where
+    Snapshot: WritableSnapshot + 'static,
+    PreviousStage: StageAPI<Snapshot>,
+{
+    type OutputIterator = WrittenRowsIterator;
+
+    fn into_iterator(
```

Comment:
I feel like we don't normally have _so_ much logic in the `into_iterator` of a stage executor? This feels like an anti-pattern/missing abstraction/not matching existing patterns?

Reply:
I think it has the same responsibilities as the other `into_iterator`s. 
I also don't know if we're going to keep the `PipelineStage`s or just push everything into the `PatternExecutor`, since we've already got most of the read stages in there anyway.

Reply:
Hmm ok maybe you can modularise this into some sub-functions so it's not so monolithic and huge? eg the way `InsertStageExecutor` does it?